File: src/ssh_manager/services/rotator.py
"""Zero-downtime, staged, single-old-archive rotation (invariant 7).

rotate: stage a replacement in ``.staging/`` → deploy the staged pubkey to every
target (the current key stays active) → verify login on each → commit ONLY after
all verify (purge any existing ``/old/`` predecessor, archive the current key
under the identical filename, promote staged, revoke old, reset inventory+audit).
On any failure before commit the staged key is discarded and the active key is
untouched. rollback is the symmetric reverse move of the single ``/old/`` key.
"""
from __future__ import annotations

import contextlib
import shutil
from dataclasses import dataclass, field
from pathlib import Path

from ..core.inventory import (
    Deployment,
    Inventory,
    KeyRecord,
    compute_expiry,
    today,
)
from ..core.manifest import Host, Manifest
from ..platforms.base import Platform
from ..providers.base import Target
from ..providers.registry import resolve as resolve_provider
from ..util import fs, log, net, perms
from ..util.errors import SshManagerError
from ..util.paths import Paths
from .keystore import KeyStore


@dataclass
class TargetResult:
    alias: str
    provider: str
    deployed: bool = False
    verified: bool = False
    revoked: bool = False


@dataclass
class RotateReport:
    key_name: str
    old_fingerprint: str = ""
    new_fingerprint: str = ""
    committed: bool = False
    message: str = ""
    targets: list[TargetResult] = field(default_factory=list)

    def format(self) -> str:
        head = "rotated" if self.committed else "rotation ABORTED"
        lines = [f"{head}: {self.key_name}"]
        if self.old_fingerprint:
            lines.append(f"  old: {self.old_fingerprint}")
        if self.new_fingerprint:
            lines.append(f"  new: {self.new_fingerprint}")
        for t in self.targets:
            lines.append(
                f"  {t.alias} ({t.provider}): "
                f"deploy={'ok' if t.deployed else 'FAIL'} "
                f"verify={'ok' if t.verified else 'no'} "
                f"revoke={'ok' if t.revoked else '-'}"
            )
        if self.message:
            lines.append(f"  {self.message}")
        return "\n".join(lines)


class Rotator:
    def __init__(self, platform: Platform, paths: Paths,
                 manifest: Manifest, inventory: Inventory) -> None:
        self._platform = platform
        self._paths = paths
        self._manifest = manifest
        self._inventory = inventory
        self._keystore = KeyStore(platform)

    # path helpers
    def _profile_and_hosts(self, key_name: str) -> tuple[str, list[Host]]:
        matches = [rk for rk in self._manifest.iter_resolved() if rk.key_name == key_name]
        if not matches:
            raise SshManagerError(f"no host in the manifest uses key {key_name!r}")
        return matches[0].profile, [m.host for m in matches]

    def _dir(self, profile: str) -> Path:
        return self._paths.ssh_dir / "profiles" / profile

    def _unreachable_targets(self, hosts: list[Host]) -> list[net.NetStatus]:
        """Reachability of every SSH-to-host target (skips API/web-panel providers,
        whose 'host' is a service endpoint, not the box itself)."""
        out: list[net.NetStatus] = []
        for host in hosts:
            provider = resolve_provider(host.provider, self._paths.providers)
            if provider.category != "server":
                continue
            st = net.check(host.hostname, host.port, ssh=True,
                           requires_vpn=host.requires_vpn, vpn_name=host.vpn_name,
                           vpn_url=host.vpn_url)
            if not st.reachable:
                out.append(st)
        return out

    # rotate
    def rotate(self, key_name: str, *, allow_unverified: bool = False,
               passphrase: str = "") -> RotateReport:
        profile, hosts = self._profile_and_hosts(key_name)
        pdir = self._dir(profile)
        cur_priv, cur_pub = pdir / key_name, pdir / f"{key_name}.pub"
        if not cur_priv.exists():
            raise SshManagerError(f"key not present: {cur_priv} - run `sshmgr reconcile` first")

        report = RotateReport(key_name=key_name)
        report.old_fingerprint = self._keystore.fingerprint(cur_pub)
        old_pub_text = cur_pub.read_text(encoding="utf-8")

        # 0. Preflight reachability: every SSH target must answer before we stage or
        # deploy anything - so an unreachable / VPN-gated host fails fast with an
        # actionable message instead of hanging mid-rotation.
        unreachable = self._unreachable_targets(hosts)
        if unreachable:
            report.message = "cannot rotate - " + "; ".join(s.message for s in unreachable)
            log.audit(self._paths.audit_log, "rotate.unreachable", key=key_name)
            return report

        # 1. Stage a fresh keypair (discard any leftover from a crashed rotation).
        staging = pdir / ".staging"
        if staging.exists():
            shutil.rmtree(staging)
        fs.ensure_dir(staging, perms.DIR_MODE)
        staged_priv = staging / key_name
        comment = f"{profile}/{hosts[0].alias} {today()}"
        gen = self._keystore.generate(
            staged_priv, key_type=self._manifest.defaults.key_type, comment=comment,
            passphrase=passphrase,
        )
        report.new_fingerprint = gen.fingerprint
        staged_pub = staging / f"{key_name}.pub"

        # 2. Deploy the staged pubkey to every target (current key still active).
        # 3. Verify login with the staged key on each target.
        results: list[TargetResult] = []
        for host in hosts:
            provider = resolve_provider(host.provider, self._paths.providers)
            tr = TargetResult(alias=host.alias, provider=provider.name)
            tgt = Target(
                alias=host.alias, hostname=host.hostname, user=host.user, port=host.port,
                pubkey_path=staged_pub, pubkey_text=staged_pub.read_text(encoding="utf-8"),
                token_env=host.token_env, identity_path=staged_priv,
                known_hosts=pdir / "known_hosts",
            )
            outcome = provider.deploy(tgt)
            # "deploy ok" = an automated deploy that succeeded, OR a manual target
            # (the user will paste it). This is what --allow-unverified accepts.
            tr.deployed = outcome.method == "manual" or outcome.verified
            tr.verified = provider.verify(tgt)
            results.append(tr)
        report.targets = results

        # Commit if every target verified by login, OR --allow-unverified and every
        # target's deploy was acknowledged (covers manual/web-panel targets).
        ready = all(t.verified for t in results) or (
            allow_unverified and all(t.deployed for t in results)
        )
        if not ready:
            # Best-effort: pull the staged pubkey back off every target where the
            # deploy already landed, so an aborted rotation doesn't leave orphan
            # keys accumulating in authorized_keys / provider accounts.
            staged_text = staged_pub.read_text(encoding="utf-8")
            for host, tr in zip(hosts, results, strict=True):
                if not tr.deployed:
                    continue
                provider = resolve_provider(host.provider, self._paths.providers)
                with contextlib.suppress(Exception):
                    provider.remove(Target(
                        alias=host.alias, hostname=host.hostname, user=host.user,
                        port=host.port, pubkey_path=staged_pub, pubkey_text=staged_text,
                        token_env=host.token_env, known_hosts=pdir / "known_hosts",
                    ))
            shutil.rmtree(staging)        # 5. discard staged; active key untouched
            report.message = ("verification failed on one or more targets - staged key "
                              "discarded (and pulled back from any target it reached), "
                              "active key untouched. (Use --allow-unverified to accept "
                              "manual/web-panel targets.)")
            log.audit(self._paths.audit_log, "rotate.abort",
                      key=key_name, old=report.old_fingerprint)
            return report

        # 4. COMMIT (only after all verified).
        old_dir = pdir / "old"
        fs.ensure_dir(old_dir, perms.DIR_MODE)
        # purge any existing predecessor for this key (enforces ≤1-old)
        (old_dir / key_name).unlink(missing_ok=True)
        (old_dir / f"{key_name}.pub").unlink(missing_ok=True)
        # archive current under the identical filename
        cur_priv.replace(old_dir / key_name)
        cur_pub.replace(old_dir / f"{key_name}.pub")
        # promote staged to canonical
        staged_priv.replace(cur_priv)
        staged_pub.replace(cur_pub)
        self._platform.set_perms(cur_priv, perms.PRIVATE_KEY_MODE)
        self._platform.set_perms(cur_pub, perms.PUBLIC_KEY_MODE)
        self._platform.set_perms(old_dir / key_name, perms.PRIVATE_KEY_MODE)
        self._platform.set_perms(old_dir / f"{key_name}.pub", perms.PUBLIC_KEY_MODE)
        shutil.rmtree(staging)

        # revoke the old public key from each target (best-effort)
        for host, tr in zip(hosts, results, strict=True):
            provider = resolve_provider(host.provider, self._paths.providers)
            old_tgt = Target(
                alias=host.alias, hostname=host.hostname, user=host.user, port=host.port,
                pubkey_path=old_dir / f"{key_name}.pub", pubkey_text=old_pub_text,
                token_env=host.token_env, known_hosts=pdir / "known_hosts",
            )
            tr.revoked = provider.remove(old_tgt)

        self._update_inventory(profile, key_name, report, results)
        report.committed = True
        log.audit(self._paths.audit_log, "rotate",
                  key=key_name, old=report.old_fingerprint, new=report.new_fingerprint)
        return report

    def _update_inventory(self, profile: str, key_name: str,
                          report: RotateReport, results: list[TargetResult]) -> None:
        ident = str(self._manifest.identity_file(profile, key_name))
        old_ident = f"~/.ssh/profiles/{profile}/old/{key_name}"
        # Drop any stale record still pointing at the single /old/ slot (a prior
        # rotation's predecessor, now purged from disk) so records don't pile up.
        for fp in [f for f, r in self._inventory.keys.items()
                   if r.path == old_ident and f != report.old_fingerprint]:
            del self._inventory.keys[fp]
        # Keep the outgoing record but mark it archived (path -> old/), so rollback
        # can find it and expiry skips it.
        if report.old_fingerprint in self._inventory.keys:
            self._inventory.keys[report.old_fingerprint].path = old_ident
        created = today()
        self._inventory.record(report.new_fingerprint, KeyRecord(
            profile=profile, path=ident, type=self._manifest.defaults.key_type,
            comment=f"{profile}/{key_name} {created}", created=created,
            rotate_after_days=self._manifest.defaults.rotate_after_days,
            expires_on=compute_expiry(created, self._manifest.defaults.rotate_after_days),
            deployments=[
                Deployment(target=t.alias, method=t.provider, date=created,
                           verified=t.verified)
                for t in results
            ],
        ))

    # rollback
    def rollback(self, key_name: str) -> RotateReport:
        profile, hosts = self._profile_and_hosts(key_name)
        pdir = self._dir(profile)
        old_priv, old_pub = pdir / "old" / key_name, pdir / "old" / f"{key_name}.pub"
        if not old_priv.exists():
            raise SshManagerError(f"no /old/ predecessor to roll back to for {key_name!r}")
        cur_priv, cur_pub = pdir / key_name, pdir / f"{key_name}.pub"

        report = RotateReport(key_name=key_name)
        report.old_fingerprint = (
            self._keystore.fingerprint(cur_pub) if cur_pub.exists() else ""
        )
        cur_pub_text = cur_pub.read_text(encoding="utf-8") if cur_pub.exists() else ""
        report.new_fingerprint = self._keystore.fingerprint(old_pub)  # the restored key

        # Plain reverse move: predecessor -> canonical (replacing current).
        cur_priv.unlink(missing_ok=True)
        cur_pub.unlink(missing_ok=True)
        old_priv.replace(cur_priv)
        old_pub.replace(cur_pub)
        self._platform.set_perms(cur_priv, perms.PRIVATE_KEY_MODE)
        self._platform.set_perms(cur_pub, perms.PUBLIC_KEY_MODE)

        # Re-deploy the restored key + revoke the (removed) rotated-in key.
        results: list[TargetResult] = []
        for host in hosts:
            provider = resolve_provider(host.provider, self._paths.providers)
            tr = TargetResult(alias=host.alias, provider=provider.name)
            # The local restore above is the part that matters; the network
            # re-deploy/revoke is best-effort, so skip (don't hang on) an
            # unreachable host rather than aborting the rollback.
            if provider.category == "server" and not net.check(
                    host.hostname, host.port, ssh=True,
                    requires_vpn=host.requires_vpn, vpn_name=host.vpn_name).reachable:
                results.append(tr)
                continue
            restored = Target(
                alias=host.alias, hostname=host.hostname, user=host.user, port=host.port,
                pubkey_path=cur_pub, pubkey_text=cur_pub.read_text(encoding="utf-8"),
                token_env=host.token_env, identity_path=cur_priv,
                known_hosts=pdir / "known_hosts",
            )
            out = provider.deploy(restored)
            # Same semantics as rotate: "deployed" == an automated deploy that
            # succeeded OR a manual target (the user pastes it).
            tr.deployed = out.method == "manual" or out.verified
            tr.verified = provider.verify(restored)
            if cur_pub_text:
                removed = Target(
                    alias=host.alias, hostname=host.hostname, user=host.user,
                    port=host.port, pubkey_path=cur_pub, pubkey_text=cur_pub_text,
                    token_env=host.token_env, known_hosts=pdir / "known_hosts",
                )
                tr.revoked = provider.remove(removed)
            results.append(tr)
        report.targets = results

        # Inventory: drop the rotated-in record; restore the predecessor to canonical.
        if report.old_fingerprint:
            self._inventory.keys.pop(report.old_fingerprint, None)
        ident = str(self._manifest.identity_file(profile, key_name))
        if report.new_fingerprint in self._inventory.keys:
            self._inventory.keys[report.new_fingerprint].path = ident
            self._inventory.keys[report.new_fingerprint].deployments = [
                Deployment(target=t.alias, method=t.provider, date=today(),
                           verified=t.verified)
                for t in results
            ]
        report.committed = True
        log.audit(self._paths.audit_log, "rollback",
                  key=key_name, restored=report.new_fingerprint)
        return report

File: internal/services/rotator/rotator.go
// Package rotator does zero-downtime, staged, single-old-archive key rotation,
// ported from services/rotator.py. rotate stages a replacement, deploys it to
// every target (the current key stays active), verifies, and only then commits
// (archive current under /old/, promote staged, revoke old). On any pre-commit
// failure the staged key is discarded and the active key is untouched. rollback is
// the symmetric reverse move of the single /old/ key.
package rotator

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/simtabi/ssh-manager/internal/core/inventory"
	"github.com/simtabi/ssh-manager/internal/core/manifest"
	"github.com/simtabi/ssh-manager/internal/core/providers"
	"github.com/simtabi/ssh-manager/internal/services/keystore"
	"github.com/simtabi/ssh-manager/internal/util/fs"
	"github.com/simtabi/ssh-manager/internal/util/log"
	"github.com/simtabi/ssh-manager/internal/util/netcheck"
	"github.com/simtabi/ssh-manager/internal/util/paths"
	"github.com/simtabi/ssh-manager/internal/util/perms"
)

// TargetResult is one target's rotation outcome.
type TargetResult struct {
	Alias    string
	Provider string
	Deployed bool
	Verified bool
	Revoked  bool
}

// RotateReport summarizes a rotate/rollback run.
type RotateReport struct {
	KeyName        string
	OldFingerprint string
	NewFingerprint string
	Committed      bool
	Message        string
	Targets        []TargetResult
}

// Format renders the rotation summary (mirrors RotateReport.format).
func (r RotateReport) Format() string {
	head := "rotation ABORTED"
	if r.Committed {
		head = "rotated"
	}
	lines := []string{head + ": " + r.KeyName}
	if r.OldFingerprint != "" {
		lines = append(lines, "  old: "+r.OldFingerprint)
	}
	if r.NewFingerprint != "" {
		lines = append(lines, "  new: "+r.NewFingerprint)
	}
	for _, t := range r.Targets {
		lines = append(lines, fmt.Sprintf("  %s (%s): deploy=%s verify=%s revoke=%s",
			t.Alias, t.Provider, okFail(t.Deployed), yesNo(t.Verified), revokeFlag(t.Revoked)))
	}
	if r.Message != "" {
		lines = append(lines, "  "+r.Message)
	}
	return strings.Join(lines, "\n")
}

func okFail(b bool) string {
	if b {
		return "ok"
	}
	return "FAIL"
}
func yesNo(b bool) string {
	if b {
		return "ok"
	}
	return "no"
}
func revokeFlag(b bool) string {
	if b {
		return "ok"
	}
	return "-"
}

// Rotator rotates keys.
type Rotator struct {
	p   paths.Paths
	m   *manifest.Manifest
	inv *inventory.Inventory
	ks  *keystore.KeyStore
}

// New builds a Rotator.
func New(p paths.Paths, m *manifest.Manifest, inv *inventory.Inventory) *Rotator {
	return &Rotator{p: p, m: m, inv: inv, ks: keystore.New()}
}

// profileAndHosts resolves a key selector to its owning profile and the hosts
// that use it. Hosts are scoped to that one profile: rotating touches files in a
// single profile directory, so pulling in a same-named key from another profile
// would deploy against hosts whose key was never rotated.
func (r *Rotator) profileAndHosts(selector string) (string, string, []manifest.Host, error) {
	ref, err := r.m.ResolveKeySelector(selector)
	if err != nil {
		return "", "", nil, err
	}
	hosts, err := r.m.HostsForKey(ref)
	if err != nil {
		return "", "", nil, err
	}
	if len(hosts) == 0 {
		return "", "", nil, fmt.Errorf("no host in the manifest uses key %q", ref)
	}
	return ref.Profile, ref.KeyName, hosts, nil
}

func (r *Rotator) dir(profile string) string {
	return filepath.Join(r.p.SSHDir, "profiles", profile)
}

func (r *Rotator) provider(h manifest.Host) providers.Provider {
	return providers.Resolve(deref(h.Provider), r.p.Providers())
}

func (r *Rotator) target(h manifest.Host, profile, pubPath, pubText, identPath string) providers.Target {
	return providers.Target{
		Alias: h.Alias, Hostname: h.Hostname, User: h.User, Port: h.Port,
		PubkeyPath: pubPath, PubkeyText: pubText, TokenEnv: deref(h.TokenEnv),
		IdentityPath: identPath, KnownHosts: filepath.Join(r.p.SSHDir, "known_hosts"),
	}
}

func (r *Rotator) unreachable(hosts []manifest.Host) []string {
	var msgs []string
	for _, h := range hosts {
		if r.provider(h).Category() != "server" {
			continue
		}
		st := netcheck.Check(h.Hostname, h.Port, h.RequiresVPN, deref(h.VPNName), deref(h.VPNURL), 4*time.Second, true)
		if !st.Reachable {
			msgs = append(msgs, st.Message())
		}
	}
	return msgs
}

// Rotate stages a fresh key, deploys+verifies it on every target, and commits only
// if all verified (or allowUnverified and all deployed). Mirrors Rotator.rotate.
func (r *Rotator) Rotate(selector string, allowUnverified bool, passphrase string) (RotateReport, error) {
	profile, keyName, hosts, err := r.profileAndHosts(selector)
	if err != nil {
		return RotateReport{}, err
	}
	pdir := r.dir(profile)
	curPriv := filepath.Join(pdir, keyName)
	curPub := curPriv + ".pub"
	if !exists(curPriv) {
		return RotateReport{}, fmt.Errorf("key not present: %s - run `sshmgr reconcile` first", curPriv)
	}
	report := RotateReport{KeyName: keyName}
	if fp, err := r.ks.Fingerprint(curPub); err == nil {
		report.OldFingerprint = fp
	}
	oldPubText := readFile(curPub)

	// 0. Preflight: every SSH target must answer before we stage/deploy.
	if msgs := r.unreachable(hosts); len(msgs) > 0 {
		report.Message = "cannot rotate - " + strings.Join(msgs, "; ")
		log.Audit(r.p.AuditLog(), "rotate.unreachable", log.Field{Key: "key", Value: keyName})
		return report, nil
	}

	// 1. Stage a fresh keypair (discard any crashed-rotation leftover).
	staging := filepath.Join(pdir, ".staging")
	_ = os.RemoveAll(staging)
	if err := fs.EnsureDir(staging, perms.DirMode); err != nil {
		return RotateReport{}, err
	}
	stagedPriv := filepath.Join(staging, keyName)
	stagedPub := stagedPriv + ".pub"
	comment := fmt.Sprintf("%s/%s %s", profile, hosts[0].Alias, inventory.Today())
	gen, err := r.ks.Generate(stagedPriv, r.m.Defaults.KeyType, comment, passphrase, false)
	if err != nil {
		return RotateReport{}, err
	}
	report.NewFingerprint = gen.Fingerprint

	// 2+3. Deploy the staged pubkey to every target, then verify login.
	stagedPubText := readFile(stagedPub)
	var results []TargetResult
	for _, h := range hosts {
		prov := r.provider(h)
		tr := TargetResult{Alias: h.Alias, Provider: prov.Name()}
		tgt := r.target(h, profile, stagedPub, stagedPubText, stagedPriv)
		out := prov.Deploy(tgt)
		tr.Deployed = out.Method == "manual" || out.Verified
		tr.Verified = prov.Verify(tgt)
		results = append(results, tr)
	}
	report.Targets = results

	ready := allTrue(results, func(t TargetResult) bool { return t.Verified }) ||
		(allowUnverified && allTrue(results, func(t TargetResult) bool { return t.Deployed }))
	if !ready {
		// Pull the staged pubkey back off every target it reached.
		for i, h := range hosts {
			if !results[i].Deployed {
				continue
			}
			tgt := r.target(h, profile, stagedPub, stagedPubText, "")
			_ = r.provider(h).Remove(tgt)
		}
		_ = os.RemoveAll(staging)
		report.Message = "verification failed on one or more targets - staged key discarded " +
			"(and pulled back from any target it reached), active key untouched. " +
			"(Use --allow-unverified to accept manual/web-panel targets.)"
		log.Audit(r.p.AuditLog(), "rotate.abort", log.Field{Key: "key", Value: keyName}, log.Field{Key: "old", Value: report.OldFingerprint})
		return report, nil
	}

	// 4. COMMIT.
	//
	// The canonical path is the only identity ssh will offer, so it has to hold a
	// working private key at every instant. This used to be four bare renames -
	// current pair out to old/, staged pair in - and any one of them failing left
	// the canonical path empty with the live key sitting in old/ or .staging/,
	// discovered the next time the user tried to push. Nothing put it back.
	//
	// Now the outgoing pair is preserved into old/ by hard link, so the canonical
	// path still points at it while the copy is made, and the staged pair is
	// renamed *over* the canonical path - which POSIX defines as atomic. A crash
	// between any two steps leaves a complete, working key behind, and a failure
	// puts back whatever had already moved.
	oldDir := filepath.Join(pdir, "old")
	if err := fs.EnsureDir(oldDir, perms.DirMode); err != nil {
		return RotateReport{}, fmt.Errorf("could not prepare the archive directory, so the rotation "+
			"was not committed (the active key is untouched): %w", err)
	}
	oldPrivSlot := filepath.Join(oldDir, keyName)
	if err := preservePair(curPriv, oldPrivSlot); err != nil {
		return RotateReport{}, fmt.Errorf("could not archive the outgoing key, so the rotation was "+
			"not committed (the active key is untouched): %w", err)
	}
	if err := os.Rename(stagedPriv, curPriv); err != nil {
		// The canonical pair never moved; drop the archive copy we just made so
		// old/ does not claim a predecessor that was never replaced.
		removePair(oldPrivSlot)
		return RotateReport{}, fmt.Errorf("could not move the new key into place, so the rotation "+
			"was not committed (the active key is untouched): %w", err)
	}
	if err := os.Rename(stagedPub, curPub); err != nil {
		// The private half already changed. Put the preserved one back so the
		// canonical pair matches itself again.
		if rerr := os.Rename(oldPrivSlot, curPriv); rerr != nil {
			return RotateReport{}, fmt.Errorf("the rotation failed midway AND could not be undone. "+
				"%s now holds the NEW private key while %s is the OLD public key, and the "+
				"previous private key is at %s. Restore by hand, or run `sshmgr rollback %s`: "+
				"%w (undo also failed: %v)", curPriv, curPub, oldPrivSlot, keyName, err, rerr)
		}
		removePair(oldPrivSlot)
		return RotateReport{}, fmt.Errorf("could not move the new public key into place, so the "+
			"rotation was undone and the previous key is active again: %w", err)
	}
	// Only now: before the swap the archive shared an inode with the live key, so
	// setting its mode would have set the live key's too.
	_ = perms.SetPerms(curPriv, perms.PrivateKeyMode)
	_ = perms.SetPerms(curPub, perms.PublicKeyMode)
	_ = perms.SetPerms(oldPrivSlot, perms.PrivateKeyMode)
	_ = perms.SetPerms(oldPrivSlot+".pub", perms.PublicKeyMode)
	_ = os.RemoveAll(staging)

	// Revoke the old public key from each target (best-effort).
	oldPub := oldPrivSlot + ".pub"
	for i := range hosts {
		tgt := r.target(hosts[i], profile, oldPub, oldPubText, "")
		results[i].Revoked = r.provider(hosts[i]).Remove(tgt)
	}
	report.Targets = results

	r.updateInventory(profile, keyName, report, results)
	report.Committed = true
	log.Audit(r.p.AuditLog(), "rotate", log.Field{Key: "key", Value: keyName},
		log.Field{Key: "old", Value: report.OldFingerprint}, log.Field{Key: "new", Value: report.NewFingerprint})
	return report, nil
}

func (r *Rotator) updateInventory(profile, keyName string, report RotateReport, results []TargetResult) {
	ident := r.m.IdentityFile(profile, keyName)
	oldIdent := fmt.Sprintf("~/.ssh/profiles/%s/old/%s", profile, keyName)
	// Drop any stale record still at the single /old/ slot (a prior predecessor).
	for fp, rec := range r.inv.Keys {
		if rec.Path == oldIdent && fp != report.OldFingerprint {
			delete(r.inv.Keys, fp)
		}
	}
	// Keep the outgoing record but mark it archived (path -> old/).
	if rec, ok := r.inv.Keys[report.OldFingerprint]; ok {
		rec.Path = oldIdent
		r.inv.Keys[report.OldFingerprint] = rec
	}
	created := inventory.Today()
	exp, _ := inventory.ComputeExpiry(created, r.m.Defaults.RotateAfterDays)
	comment := fmt.Sprintf("%s/%s %s", profile, keyName, created)
	r.inv.Record(report.NewFingerprint, inventory.KeyRecord{
		Profile: profile, Path: ident, Type: r.m.Defaults.KeyType,
		Comment: &comment, Created: &created, RotateAfterDays: r.m.Defaults.RotateAfterDays,
		ExpiresOn: &exp, Deployments: deployments(results, created),
	})
}

// Rollback restores the single /old/ predecessor (plain reverse move), re-deploys
// it, and revokes the rotated-in key. Mirrors Rotator.rollback.
func (r *Rotator) Rollback(selector string) (RotateReport, error) {
	profile, keyName, hosts, err := r.profileAndHosts(selector)
	if err != nil {
		return RotateReport{}, err
	}
	pdir := r.dir(profile)
	oldPriv := filepath.Join(pdir, "old", keyName)
	oldPub := oldPriv + ".pub"
	if !exists(oldPriv) {
		return RotateReport{}, fmt.Errorf("no /old/ predecessor to roll back to for %q", keyName)
	}
	curPriv := filepath.Join(pdir, keyName)
	curPub := curPriv + ".pub"

	report := RotateReport{KeyName: keyName}
	if exists(curPub) {
		if fp, err := r.ks.Fingerprint(curPub); err == nil {
			report.OldFingerprint = fp
		}
	}
	curPubText := readFile(curPub)
	if fp, err := r.ks.Fingerprint(oldPub); err == nil {
		report.NewFingerprint = fp
	}

	// Reverse move: predecessor -> canonical, replacing the rotated-in key.
	//
	// Nothing is removed first. This used to delete the canonical pair and then
	// move the predecessor in, so a failed rename left the canonical path empty
	// with the key it had just deleted gone for good - a rollback that locks you
	// out is worse than the rotation it was undoing. Rename over an existing file
	// is atomic, so the canonical path holds a complete private key throughout.
	if err := os.Rename(oldPriv, curPriv); err != nil {
		return RotateReport{}, fmt.Errorf("could not restore the previous private key, so nothing "+
			"was rolled back (the rotated-in key is still active): %w", err)
	}
	if err := os.Rename(oldPub, curPub); err != nil {
		// The private half is restored and working; only the .pub is stale. It is
		// derivable from the private key, so derive it rather than leave a pair
		// that does not match itself.
		if !r.rederivePub(curPriv, curPub) {
			return RotateReport{}, fmt.Errorf("the previous private key was restored to %s and is "+
				"active, but its public key could not be put back: %s still holds the rotated-in "+
				"public key. Re-derive it with `ssh-keygen -y -f %s > %s`: %w",
				curPriv, curPub, curPriv, curPub, err)
		}
		log.Audit(r.p.AuditLog(), "rollback.pub-rederived", log.Field{Key: "key", Value: keyName})
	}
	_ = perms.SetPerms(curPriv, perms.PrivateKeyMode)
	_ = perms.SetPerms(curPub, perms.PublicKeyMode)

	restoredPubText := readFile(curPub)
	var results []TargetResult
	for _, h := range hosts {
		prov := r.provider(h)
		tr := TargetResult{Alias: h.Alias, Provider: prov.Name()}
		if prov.Category() == "server" {
			st := netcheck.Check(h.Hostname, h.Port, h.RequiresVPN, deref(h.VPNName), "", 4*time.Second, true)
			if !st.Reachable {
				results = append(results, tr)
				continue
			}
		}
		restored := r.target(h, profile, curPub, restoredPubText, curPriv)
		out := prov.Deploy(restored)
		tr.Deployed = out.Method == "manual" || out.Verified
		tr.Verified = prov.Verify(restored)
		if curPubText != "" {
			removed := r.target(h, profile, curPub, curPubText, "")
			tr.Revoked = prov.Remove(removed)
		}
		results = append(results, tr)
	}
	report.Targets = results

	if report.OldFingerprint != "" {
		delete(r.inv.Keys, report.OldFingerprint)
	}
	ident := r.m.IdentityFile(profile, keyName)
	if rec, ok := r.inv.Keys[report.NewFingerprint]; ok {
		rec.Path = ident
		rec.Deployments = deployments(results, inventory.Today())
		r.inv.Keys[report.NewFingerprint] = rec
	}
	report.Committed = true
	log.Audit(r.p.AuditLog(), "rollback", log.Field{Key: "key", Value: keyName}, log.Field{Key: "restored", Value: report.NewFingerprint})
	return report, nil
}

// preservePair copies the keypair at src to dst without disturbing src.
//
// A hard link, not a copy: the two names share one inode, so the source keeps
// working while the archive exists and no private key bytes are written a second
// time. src and dst are both inside one profile directory, so a link is
// essentially always possible; a filesystem that refuses one (a bind mount over
// old/, say) falls back to copying, since preserving the outgoing key matters
// more than how it is preserved.
func preservePair(srcPriv, dstPriv string) error {
	for _, ext := range []string{"", ".pub"} {
		src, dst := srcPriv+ext, dstPriv+ext
		if !exists(src) {
			continue // a half pair is the caller's problem, not this one's
		}
		_ = os.Remove(dst) // the single archive slot holds one predecessor
		if err := os.Link(src, dst); err == nil {
			continue
		}
		if err := copyFile(src, dst); err != nil {
			return err
		}
	}
	return nil
}

// copyFile is the fallback for preservePair. Private key modes are re-asserted
// by the caller's perms sweep; 0600 here means the file is never briefly
// readable by anyone else.
func copyFile(src, dst string) error {
	body, err := os.ReadFile(src)
	if err != nil {
		return err
	}
	return fs.WriteTextAtomic(dst, string(body), perms.PrivateKeyMode)
}

// removePair deletes both halves of a keypair, ignoring absent files.
func removePair(priv string) {
	_ = os.Remove(priv)
	_ = os.Remove(priv + ".pub")
}

// rederivePub regenerates a public key file from its private key. Reports
// whether it succeeded - an encrypted key cannot be derived without its
// passphrase, and the caller has to say so rather than pretend.
func (r *Rotator) rederivePub(priv, pub string) bool {
	derived, _, err := r.ks.PublicFromPrivate(priv)
	if err != nil || derived == "" {
		return false
	}
	return fs.WriteTextAtomic(pub, derived+"\n", perms.PublicKeyMode) == nil
}

func deployments(results []TargetResult, date string) []inventory.Deployment {
	out := make([]inventory.Deployment, 0, len(results))
	d := date
	for _, t := range results {
		dd := d
		out = append(out, inventory.Deployment{Target: t.Alias, Method: t.Provider, Date: &dd, Verified: t.Verified})
	}
	return out
}

func allTrue(rs []TargetResult, pred func(TargetResult) bool) bool {
	for _, t := range rs {
		if !pred(t) {
			return false
		}
	}
	return true
}

func exists(p string) bool { _, err := os.Stat(p); return err == nil }
func readFile(p string) string {
	b, err := os.ReadFile(p)
	if err != nil {
		return ""
	}
	return string(b)
}
func deref(s *string) string {
	if s == nil {
		return ""
	}
	return *s
}
